### ISSUE TITLE
Unexpected memory usage for efficientnetb1?

### ISSUE THREAD

**Issue by samedii** [tensorflow/neural-structured-learning#22]
I can train much larger models than efficientnetb1 without nsl (moreover, only the decoder is trainable here) without using much RAM. NSL's memory usage seems to be growing without end. Is this expected behaviour to use 13+ GB of RAM for this?

Edit: Using `nsl.keras.AdversarialRegularization`

Using this implementation: https://github.com/qubvel/segmentation_models/tree/master/segmentation_models

![image](https://user-images.githubusercontent.com/1872050/65919737-8333d300-e3dd-11e9-8a57-aec678f6fc72.png)

Edit: Uses 2.6 GB of RAM when training with a single layer Conv2D (with adverserial model) and around 4-5 GB when running efficientnetb1 without nsl

**Comment by samedii** [comment]
Works with Unet + mobilenetv2 using 10 GB of RAM so it's probably at least not growing without bound.

**Comment by ppham27** [comment]
Memory usage will always be higher in adversarial training to some degree, but that does seem excessive. I wonder if some of the overhead is coming from the separate gradient tape. In an image model the activations are rather large, and the tape caches all intermediate activations. Ideally, common subexpression elimination should catch this, though, when the training loop is being compiled. What version of TensorFlow are you using? If you can provide a working code example, we may be able to debug further.

We don't have easy access to the tape that Keras uses. You could try using a custom training loop with a single tape.

**Comment by csferng** [comment]
@samedii, sorry for the super late reply.

I managed to reproduce the issue with EfficientNet using TensorFlow 2.0. As @ppham27 mentioned, the intermediate activations for the adversarial examples did cost some extra memory. Also Keras model subclassing (which `AdversarialRegularization` does) seems to be taking some space as well.

Thankfully the situation is much improved with the recent TensorFlow 2.1-rc0 release. In my experiment, the adversarial regularized model now takes 4G of RAM when using TensorFlow 2.1-rc0, comparing to 12G using TensorFlow 2.0. The base EfficientNet model takes 3G of RAM regardless of TensorFlow versions.

I am closing this issue for now. Feel free to reopen it or create a new one if you need anything else.
